src/backend/cpu/kernel/canny.hpp

```diff
@@ -0,0 +1,237 @@
+                                    mag1  = (1-alpha)*mag[offset-dims[0]] + alpha*mag[offset-dims[0]+1];
+                                    mag2  = (1-alpha)*mag[offset+dims[0]] + alpha*mag[offset+dims[0]-1];
+                                }
+                            }
+                        } else {
+                            if (dY[offset]>=0) {
+                                if ((dX[offset]+dY[offset]) >= 0) {
+                                    // South of South West
+                                    alpha = (float)-dX[offset]/dY[offset];
+                                    mag1  = (1-alpha)*mag[offset+dims[0]] + alpha*mag[offset+dims[0]-1];
+                                    mag2  = (1-alpha)*mag[offset-dims[0]] + alpha*mag[offset-dims[0]+1];
+                                } else {
+                                    // West of South West
+                                    alpha = (float)dY[offset]/-dX[offset];
+                                    mag1  = (1-alpha)*mag[offset-1] + alpha*mag[offset+dims[0]-1];
+                                    mag2  = (1-alpha)*mag[offset+1] + alpha*mag[offset-dims[0]+1];
+                                }
+                            } else {
+                                if ((-dX[offset]+dY[offset]) >= 0) {
+                                    // West of North West
+                                    alpha = (float)-dY[offset]/dX[offset];
+                                    mag1  = (1-alpha)*mag[offset-1] + alpha*mag[offset-dims[0]-1];
+                                    mag2  = (1-alpha)*mag[offset+1] + alpha*mag[offset+dims[0]+1];
+                                } else {
+                                    // North of North West
+                                    alpha = (float)dX[offset]/-dY[offset];
+                                    mag1  = (1-alpha)*mag[offset-dims[0]] + alpha*mag[offset-dims[0]-1];
+                                    mag2  = (1-alpha)*mag[offset+dims[0]] + alpha*mag[offset+dims[0]+1];
+                                }
+                            }
+                        }
+
+                        if (mag[offset]>mag1 && mag[offset]>mag2) {
+                            out[offset] = mag[offset] > 255 ? EDGE : (unsigned char)mag[offset];
+                        } else {
+                            out[offset] = SUPPRESS;
+                        }
+                    }
+                }
+            }
+
+            out += strides[2];
+            mag += strides[2];
+            dX  += strides[2];
+            dY  += strides[2];
+        }
+        out += strides[3];
+        mag += strides[3];
+        dX  += strides[3];
+        dY  += strides[3];
+    }
+}
+
+template<typename T>
+void traceEdge(T* out, const T* strong, const T* weak, int t, int width)
+{
+    if (!out || !strong || !weak)
+        return;
+
+    const T EDGE = 1;
+
+    std::list<dim_t> edges; // list of edges to be checked
```

Comment:
I intended to peek and pop from front but didn't notice i wrote back. I will change that to front now.